{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "GenFigS22Data.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/pachterlab/GRNP_2020/blob/master/notebooks/figure_generation/GenFigS22Data.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dhydD96df06z"
      },
      "source": [
        "**Precalculates data for figure S22**\n",
        "\n",
        "This notebook precalculates the data for supplementary figure 22, since there are some heavy calculation steps involved for generating the figure. The code runs experiments on simulated data. This notebook may take a few hours to run. The details of what the code does can be found in supplementary note 2 of the paper.\n",
        "\n",
        "Steps:\n",
        "1. Download the code\n",
        "2. Setup the R environment\n",
        "4. Run the calculations and save the data to disk\n",
        "\n",
        "The data for this figure is produced by the following notebooks:\n",
        "\n",
        "Processing of FASTQ files with kallisto and bustools:\n",
        "\n",
        "https://github.com/pachterlab/GRNP_2020/blob/master/notebooks/FASTQ_processing/ProcessPBMC_V3_3.ipynb\n",
        "\n",
        "Preprocessing of BUG files:\n",
        "\n",
        "https://github.com/pachterlab/GRNP_2020/blob/master/notebooks/R_processing/ProcessR_PBMC_V3_3.ipynb"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "h8RnKVMXgbzr"
      },
      "source": [
        "**1. Download the code and processed data**"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "doUAtCxIyOiI",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d31a93cc-2711-41f2-f1f3-5e1d10c53496"
      },
      "source": [
        "#download the R code\n",
        "![ -d \"GRNP_2020\" ] && rm -r GRNP_2020\n",
        "\n",
        "!git clone https://github.com/pachterlab/GRNP_2020.git\n"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Cloning into 'GRNP_2020'...\n",
            "remote: Enumerating objects: 359, done.\u001b[K\n",
            "remote: Counting objects: 100% (359/359), done.\u001b[K\n",
            "remote: Compressing objects: 100% (299/299), done.\u001b[K\n",
            "remote: Total 2077 (delta 264), reused 87 (delta 60), pack-reused 1718\u001b[K\n",
            "Receiving objects: 100% (2077/2077), 10.92 MiB | 20.15 MiB/s, done.\n",
            "Resolving deltas: 100% (1449/1449), done.\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dUNSQ1qBZb2g",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "a329b801-bb3c-486f-c34e-1497b6a8de9a"
      },
      "source": [
        "#download processed data from Zenodo for all datasets\n",
        "![ -d \"figureData\" ] && rm -r figureData\n",
        "!mkdir figureData\n",
        "!cd figureData && wget https://zenodo.org/record/4661263/files/PBMC_V3_3.zip?download=1 && unzip 'PBMC_V3_3.zip?download=1' && rm 'PBMC_V3_3.zip?download=1'\n",
        "\n"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-04-05 12:47:48--  https://zenodo.org/record/4661263/files/PBMC_V3_3.zip?download=1\n",
            "Resolving zenodo.org (zenodo.org)... 137.138.76.77\n",
            "Connecting to zenodo.org (zenodo.org)|137.138.76.77|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1819306744 (1.7G) [application/octet-stream]\n",
            "Saving to: ‘PBMC_V3_3.zip?download=1’\n",
            "\n",
            "PBMC_V3_3.zip?downl 100%[===================>]   1.69G  8.55MB/s    in 5m 19s  \n",
            "\n",
            "2021-04-05 12:53:08 (5.44 MB/s) - ‘PBMC_V3_3.zip?download=1’ saved [1819306744/1819306744]\n",
            "\n",
            "Archive:  PBMC_V3_3.zip?download=1\n",
            "   creating: PBMC_V3_3/\n",
            "  inflating: PBMC_V3_3/Bug_10.RData  \n",
            "  inflating: PBMC_V3_3/Bug_100.RData  \n",
            "  inflating: PBMC_V3_3/Bug_20.RData  \n",
            "  inflating: PBMC_V3_3/Bug_40.RData  \n",
            "  inflating: PBMC_V3_3/Bug_5.RData   \n",
            "  inflating: PBMC_V3_3/Bug_60.RData  \n",
            "  inflating: PBMC_V3_3/Bug_80.RData  \n",
            "  inflating: PBMC_V3_3/ds_summary.txt  \n",
            "  inflating: PBMC_V3_3/pooledHist.RData  \n",
            "  inflating: PBMC_V3_3/pooledHistDS.RData  \n",
            "  inflating: PBMC_V3_3/PredEvalData.RDS  \n",
            "  inflating: PBMC_V3_3/Stats.RData   \n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "oesgTqLO0Qje",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "a13e5713-7888-4ff7-82f1-5d79e6abdf60"
      },
      "source": [
        "#Check that download worked\n",
        "!cd figureData && ls -l && cd PBMC_V3_3 && ls -l"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "total 4\n",
            "drwxr-xr-x 2 root root 4096 Jul  1  2020 PBMC_V3_3\n",
            "total 1899864\n",
            "-rw-r--r-- 1 root root 414359213 Jun 30  2020 Bug_100.RData\n",
            "-rw-r--r-- 1 root root 136422466 Jun 30  2020 Bug_10.RData\n",
            "-rw-r--r-- 1 root root 218056792 Jun 30  2020 Bug_20.RData\n",
            "-rw-r--r-- 1 root root 310517605 Jun 30  2020 Bug_40.RData\n",
            "-rw-r--r-- 1 root root  77512033 Jun 30  2020 Bug_5.RData\n",
            "-rw-r--r-- 1 root root 362521844 Jun 30  2020 Bug_60.RData\n",
            "-rw-r--r-- 1 root root 392412575 Jun 30  2020 Bug_80.RData\n",
            "-rw-r--r-- 1 root root      1060 Jul  1  2020 ds_summary.txt\n",
            "-rw-r--r-- 1 root root    272007 Jul  1  2020 pooledHistDS.RData\n",
            "-rw-r--r-- 1 root root    565680 Jul  1  2020 pooledHist.RData\n",
            "-rw-r--r-- 1 root root  30576428 Jul  1  2020 PredEvalData.RDS\n",
            "-rw-r--r-- 1 root root   2215477 Jun 30  2020 Stats.RData\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sCmhNVdYgkWH"
      },
      "source": [
        "**2. Prepare the R environment**"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5Gt6rQkSXriM"
      },
      "source": [
        "#switch to R mode\n",
        "%reload_ext rpy2.ipython\n"
      ],
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "jJ3rQJCdgeJa",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "05da0cc9-3e8c-40d9-d98e-ac78fa5bdbd9"
      },
      "source": [
        "#install the R packages and setup paths\n",
        "%%R\n",
        "#install.packages(\"qdapTools\")\n",
        "install.packages(\"dplyr\")\n",
        "install.packages(\"preseqR\")\n",
        "#install.packages(\"DescTools\")\n",
        "#install.packages(\"stringdist\")\n"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "R[write to console]: Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "R[write to console]: trying URL 'https://cran.rstudio.com/src/contrib/dplyr_1.0.5.tar.gz'\n",
            "\n",
            "R[write to console]: Content type 'application/x-gzip'\n",
            "R[write to console]:  length 949019 bytes (926 KB)\n",
            "\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: \n",
            "\n",
            "R[write to console]: downloaded 926 KB\n",
            "\n",
            "\n",
            "R[write to console]: \n",
            "\n",
            "R[write to console]: \n",
            "R[write to console]: The downloaded source packages are in\n",
            "\t‘/tmp/Rtmp6emQLH/downloaded_packages’\n",
            "R[write to console]: \n",
            "R[write to console]: \n",
            "\n",
            "R[write to console]: Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "R[write to console]: also installing the dependency ‘polynom’\n",
            "\n",
            "\n",
            "R[write to console]: trying URL 'https://cran.rstudio.com/src/contrib/polynom_1.4-0.tar.gz'\n",
            "\n",
            "R[write to console]: Content type 'application/x-gzip'\n",
            "R[write to console]:  length 243462 bytes (237 KB)\n",
            "\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: \n",
            "\n",
            "R[write to console]: downloaded 237 KB\n",
            "\n",
            "\n",
            "R[write to console]: trying URL 'https://cran.rstudio.com/src/contrib/preseqR_4.0.0.tar.gz'\n",
            "\n",
            "R[write to console]: Content type 'application/x-gzip'\n",
            "R[write to console]:  length 84783 bytes (82 KB)\n",
            "\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: =\n",
            "R[write to console]: \n",
            "\n",
            "R[write to console]: downloaded 82 KB\n",
            "\n",
            "\n",
            "R[write to console]: \n",
            "\n",
            "R[write to console]: \n",
            "R[write to console]: The downloaded source packages are in\n",
            "\t‘/tmp/Rtmp6emQLH/downloaded_packages’\n",
            "R[write to console]: \n",
            "R[write to console]: \n",
            "\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "x56fjfCSicrp"
      },
      "source": [
        "**3. Define general functions for generating synthetic data**\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "V37XLBAO68oR"
      },
      "source": [
        "#First set some path variables\n",
        "%%R\n",
        "source(\"GRNP_2020/RCode/pathsGoogleColab.R\")\n"
      ],
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "R6kuhOmzZL_X",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "fd9c14a9-4a8a-4417-87c4-2d80afcfbc75"
      },
      "source": [
        "#Import the code for prediction (available in other notebooks)\n",
        "%%R\n",
        "source(paste0(sourcePath,\"ButterflyHelpers.R\"))\n",
        "source(paste0(sourcePath,\"preseqHelpers.R\"))\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "R[write to console]: \n",
            "Attaching package: ‘dplyr’\n",
            "\n",
            "\n",
            "R[write to console]: The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    filter, lag\n",
            "\n",
            "\n",
            "R[write to console]: The following objects are masked from ‘package:base’:\n",
            "\n",
            "    intersect, setdiff, setequal, union\n",
            "\n",
            "\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "tkQeT362BM7V"
      },
      "source": [
        "####################\n",
        "#Define functions for simulated data generation\n",
        "####################\n",
        "%%R\n",
        "standardExpr = 1000 #gene expression used in most simulations\n",
        "#We generate data as negative binomial\n",
        "\n",
        "#We need to find realistic parameterizations for the negative binomials\n",
        "#We use a size of 1, which is a reasonable number when looking at highly expressed genes\n",
        "#The relation between mu and FSCM is then calculated. For each gene, we randomly select\n",
        "#a fraction of single-copy molecules (FSCM) from the histogram of PBMC_V3_3 (as shown in Fig S1).\n",
        "#That value is translated to an NB mean value using linear interpolation\n",
        "\n",
        "calcFSCM = function(mu, size=1) {\n",
        "  dnbinom(1, mu=mu, size=size) / (1-dnbinom(0, mu=mu, size=size))\n",
        "}\n",
        "\n",
        "\n",
        "means = 2^seq(-7,3,by=0.2)\n",
        "FSCMs = rep(NA, length(means))\n",
        "for (i in 1:length(means)) {\n",
        "  FSCMs[i] = calcFSCM(means[i])\n",
        "}\n",
        "\n",
        "#Get a list of FSCM values from the PBMC_V3_3 dataset, and convert them to mu\n",
        "dsid = \"PBMC_V3_3\"\n",
        "loadStats(dsid)\n",
        "statsV3 = getStats(dsid)\n",
        "FSCMsV3 = statsV3$FracOnes_PBMC_V3_3_d_100[statsV3$UMIs_PBMC_V3_3_d_100 >= 30]\n",
        "CU = statsV3$CountsPerUMI_PBMC_V3_3_d_100[statsV3$UMIs_PBMC_V3_3_d_100 >= 30]\n",
        "max(CU) # 8.840637 it makes sense to assume that mean is not larger than 8\n",
        "mus = approx(x = FSCMs, y = means, xout = FSCMsV3, method = \"linear\", yleft=min(means), yright=max(means))$y\n",
        "\n",
        "\n",
        "#returns a histogram and a vector of UMIs per cell\n",
        "genGeneData = function(numCells, geneExpr, mu, totMoleculesPerCell=12000) {\n",
        "  h = rep(NA, 1000) #histogram, skip the zeros, i.e. starts at 1\n",
        "  \n",
        "  totGeneMolPerCell = geneExpr/10^6*totMoleculesPerCell\n",
        "  numUMIs = round(totGeneMolPerCell*numCells)\n",
        "  #Generate CU histogram\n",
        "  molecules = 0\n",
        "  #don't accept zero molecules, it messes everything up in the analysis further down\n",
        "  while (sum(molecules) == 0) {\n",
        "    molecules = rnbinom(numUMIs, mu=mu, size=1)\n",
        "  }\n",
        "  molecules[molecules > 1000] = 1000 #remove outliers for practical reasons\n",
        "  seenMolecules = molecules[molecules > 0]\n",
        "  h = hist(seenMolecules, breaks=seq(0.5, 1000+0.5, by=1), plot = F)$counts\n",
        "\n",
        "  nonZeroMol = sum(h)\n",
        "  \n",
        "  #Generate counts across cells distribution\n",
        "  x = sample(numCells,nonZeroMol,replace=TRUE)\n",
        "  m = hist(x, breaks=seq(0.5, numCells+0.5, by=1), plot = F)$counts\n",
        "\n",
        "\n",
        "  return(list(m,h))\n",
        "}\n",
        "\n",
        "trimZeros = function(h) {\n",
        "  return(h[1:max( which( h != 0 ))])  \n",
        "}\n",
        "\n",
        "predZTNB = function(h, t) {\n",
        "  #Get the histogram\n",
        "  freq = 1:length(h)\n",
        "  counts = h\n",
        "  added = 0\n",
        "  #preseq cannot handle if we have only ones, so modify the histogram slightly\n",
        "  if ((length(freq)==1) & (freq[1] == 1)) {\n",
        "    added = 2\n",
        "    freq = c(1,2)\n",
        "    counts = c(counts[1]+1,1)#room for improvement here\n",
        "  }\n",
        "  dd = as.matrix(data.frame(freq,counts));\n",
        "  rSACZTNB = mod.ztnb.rSAC(dd, incTol = 1e-5, iterIncTol = 200);\n",
        "  newCountsZTNB = rSACZTNB(t)\n",
        "  newCountsZTNB[newCountsZTNB < 0] = 0\n",
        "  return(newCountsZTNB - added);\n",
        "}\n",
        "\n",
        "binomialDowns = function(h, fractionToKeep) {\n",
        "  \n",
        "  lh = length(h)\n",
        "  hd = rep(0,lh)\n",
        "  for (k in 1:lh) {\n",
        "    dens = dbinom(1:lh, k, fractionToKeep)\n",
        "    hd = hd + dens*h[k]\n",
        "  }\n",
        "  estTotCounts = sum(hd)\n",
        "}\n",
        "\n",
        "#mus can either be a single value or a vector, from which one randomly will be chosen\n",
        "#set mus2 to NA if we want to use the same mu for both conditions (can be used for depth calculations for example)\n",
        "evaluateCondition = function(geneExpr, mus1, mus2, depth1 = 1, depth2 = 1, DEExprFactor = 4, numCells = 5000, numGenes = 500, fractionDE = 0.5, useBinDowns = FALSE, minLFC = 0.25) {\n",
        "  geneExpr2 = geneExpr * DEExprFactor\n",
        "\n",
        "  print(\"Generating data...\")\n",
        "  \n",
        "  #randomize which genes should be DE - place the first half in the first group, the other half in the other, no overlaps\n",
        "  numDEDiv2 = round(numGenes*fractionDE/2) #half increases first group, half the second\n",
        "  numDE = numDEDiv2*2\n",
        "  deGenes = sample(numGenes, numDE, replace=FALSE)\n",
        "  isDE1 = 1:numGenes %in% deGenes[1:numDEDiv2]\n",
        "  isDE2 = 1:numGenes %in% deGenes[(1:numDEDiv2)+numDEDiv2]\n",
        "  isDE = isDE1 | isDE2\n",
        "  #sum(isDE) #ok\n",
        "  ge1 = rep(geneExpr, numGenes)\n",
        "  ge1[isDE1] = geneExpr2\n",
        "  ge2 = rep(geneExpr, numGenes)\n",
        "  ge2[isDE2] = geneExpr2\n",
        "  \n",
        "  #the first numCells cells is the first group, the next numCells is the other group\n",
        "  geMatrix = matrix(nrow=numGenes, ncol=numCells*2)\n",
        "  colnames(geMatrix) = as.character(1:(2*numCells))\n",
        "  rownames(geMatrix) = as.character(1:(numGenes))\n",
        "  \n",
        "  histGroup1 = matrix(nrow=numGenes, ncol=1000)\n",
        "  histGroup2 = matrix(nrow=numGenes, ncol=1000)\n",
        "  \n",
        "  meta = data.frame(group=as.factor(c(rep(0,numCells), rep(1,numCells))))\n",
        "  rownames(meta) = as.character(1:(2*numCells))\n",
        "  \n",
        "  mus1Rnd = sample(length(mus1), numGenes, replace=TRUE)\n",
        "  musr1 = mus1[mus1Rnd]*depth1\n",
        "  if (is.na(mus2[1])) {\n",
        "    print(paste0(\"Using depth: \", depth2))\n",
        "    musr2 = mus1[mus1Rnd]*depth2\n",
        "  } else {\n",
        "    musr2 = mus2[sample(length(mus2), numGenes, replace=TRUE)]*depth2\n",
        "  }\n",
        "  \n",
        "  for(i in 1:numGenes) {\n",
        "    lst = genGeneData(numCells, ge1[i], musr1[i])\n",
        "    lst2 = genGeneData(numCells, ge2[i], musr2[i])\n",
        "    geMatrix[i,1:numCells] = lst[[1]]\n",
        "    geMatrix[i,((1:numCells) + numCells)] = lst2[[1]]\n",
        "    histGroup1[i,] = lst[[2]]\n",
        "    histGroup2[i,] = lst2[[2]]\n",
        "  }\n",
        "\n",
        "\n",
        "  corrMatrix = geMatrix\n",
        "  \n",
        "  #Predict using ZTNB\n",
        "  if (!useBinDowns) {\n",
        "    print(\"Predicting...\")\n",
        "    gex1 = rowSums(geMatrix[1:numGenes,1:numCells])\n",
        "    gex2 = rowSums(geMatrix[1:numGenes,(1:numCells) + numCells])\n",
        "    gexPred1 = rep(NA, numGenes)\n",
        "    gexPred2 = rep(NA, numGenes)\n",
        "    for (i in 1:numGenes) {\n",
        "      gexPred1[i] = predZTNB(trimZeros(histGroup1[i,]), 10^20)\n",
        "      gexPred2[i] = predZTNB(trimZeros(histGroup2[i,]), 10^20)\n",
        "    }\n",
        "    \n",
        "    #scale to the same library size as before\n",
        "    gexPred1Norm = gexPred1*sum(gex1)/sum(gexPred1)\n",
        "    gexPred2Norm = gexPred2*sum(gex2)/sum(gexPred2)\n",
        "    scale1 = gexPred1Norm/gex1\n",
        "    scale2 = gexPred2Norm/gex2\n",
        "    \n",
        "    corrMatrix[1:numGenes, 1:numCells] = geMatrix[1:numGenes, 1:numCells]*scale1\n",
        "    corrMatrix[1:numGenes, (1:numCells)+numCells] = geMatrix[1:numGenes, (1:numCells)+numCells]*scale2\n",
        "  } else {\n",
        "    print(\"Running binomial downsampling...\")\n",
        "    #use the downsampling value sent in\n",
        "    depthChange = depth2/depth1\n",
        "    if (depthChange > 1) { # the second group should be downsampled\n",
        "      histGroup = histGroup2\n",
        "      frac = 1/depthChange\n",
        "    } else {\n",
        "      histGroup = histGroup1\n",
        "      frac = depthChange\n",
        "    }\n",
        "    \n",
        "    \n",
        "    gex1 = rowSums(geMatrix[1:numGenes,1:numCells])\n",
        "    gex2 = rowSums(geMatrix[1:numGenes,(1:numCells) + numCells])\n",
        "    gexPred = rep(NA, numGenes)\n",
        "    for (i in 1:numGenes) {\n",
        "      gexPred[i] = binomialDowns(trimZeros(histGroup[i,]), frac)\n",
        "    }\n",
        "    \n",
        "    #scale to the same library size as before\n",
        "    if (depthChange > 1) {\n",
        "      gexPred2Norm = gexPred*sum(gex1)/sum(gexPred)\n",
        "      scale2 = gexPred2Norm/gex2\n",
        "      corrMatrix[1:numGenes, (1:numCells)+numCells] = geMatrix[1:numGenes, (1:numCells)+numCells]*scale2\n",
        "    } else {\n",
        "      gexPred1Norm = gexPred*sum(gex2)/sum(gexPred)\n",
        "      scale1 = gexPred1Norm/gex1\n",
        "      corrMatrix[1:numGenes, 1:numCells] = geMatrix[1:numGenes, 1:numCells]*scale1\n",
        "    }\n",
        "  }\n",
        "\n",
        "  #do library size normalization of the two groups - this matters for example when you have different read depths\n",
        "  secFactor = sum(geMatrix[1:numGenes, 1:numCells])/sum(geMatrix[1:numGenes, (1:numCells)+numCells])\n",
        "  geMatrix[1:numGenes, (1:numCells)+numCells] = geMatrix[1:numGenes, (1:numCells)+numCells] * secFactor\n",
        "  secFactor = sum(corrMatrix[1:numGenes, 1:numCells])/sum(corrMatrix[1:numGenes, (1:numCells)+numCells])\n",
        "  corrMatrix[1:numGenes, (1:numCells)+numCells] = corrMatrix[1:numGenes, (1:numCells)+numCells] * secFactor\n",
        "  \n",
        "  \n",
        "\n",
        "  #extract fold changes for false positives and false negatives\n",
        "  lfcUnc = log2(rowSums(geMatrix[1:numGenes,1:numCells])/rowSums(geMatrix[1:numGenes,((1:numCells) + numCells)]))\n",
        "  \n",
        "\n",
        "  #extract fold changes for false positives and false negatives\n",
        "  lfcCorr = log2(rowSums(corrMatrix[1:numGenes,1:numCells])/rowSums(corrMatrix[1:numGenes,((1:numCells) + numCells)]))\n",
        "\n",
        "  return(list(isDE, lfcUnc, lfcCorr))\n",
        "}\n",
        "\n"
      ],
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YkQ3tCRUF04Z",
        "outputId": "4d1187e8-1bde-49bb-c0ee-9c535b2ab7eb"
      },
      "source": [
        "%%R\n",
        "set.seed(1)\n",
        "#############################\n",
        "# Depth\n",
        "#############################\n",
        "\n",
        "#We set the depth2 parameter\n",
        "depths = 2^((-5):5) #5 is about the limit before the histograms starts to get truncated because of the limited length of the histograms\n",
        "depthDE = list()\n",
        "depthFCUnc = list()\n",
        "depthFCCorr = list()\n",
        "\n",
        "for(i in 1:length(depths)) {\n",
        "  print(paste0(i, \" of \", length(depths)))\n",
        "  res = evaluateCondition(standardExpr, mus1=mus, mus2=NA, depth2 = depths[i], useBinDowns = TRUE, DEExprFactor = 2, numGenes=5000)\n",
        "  depthDE = c(depthDE, list(res[[1]]))\n",
        "  depthFCUnc = c(depthFCUnc, list(res[[2]]))\n",
        "  depthFCCorr = c(depthFCCorr, list(res[[3]]))\n",
        "}\n",
        "\n",
        "depthData = list(de = depthDE, \n",
        "                 fcUnc = depthFCUnc,\n",
        "                 fcCorr = depthFCCorr)\n",
        "saveRDS(depthData, file =paste0(figure_data_path, \"simDepthData.RDS\"))"
      ],
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[1] \"1 of 11\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Using depth: 0.03125\"\n",
            "[1] \"Running binomial downsampling...\"\n",
            "[1] \"2 of 11\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Using depth: 0.0625\"\n",
            "[1] \"Running binomial downsampling...\"\n",
            "[1] \"3 of 11\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Using depth: 0.125\"\n",
            "[1] \"Running binomial downsampling...\"\n",
            "[1] \"4 of 11\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Using depth: 0.25\"\n",
            "[1] \"Running binomial downsampling...\"\n",
            "[1] \"5 of 11\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Using depth: 0.5\"\n",
            "[1] \"Running binomial downsampling...\"\n",
            "[1] \"6 of 11\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Using depth: 1\"\n",
            "[1] \"Running binomial downsampling...\"\n",
            "[1] \"7 of 11\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Using depth: 2\"\n",
            "[1] \"Running binomial downsampling...\"\n",
            "[1] \"8 of 11\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Using depth: 4\"\n",
            "[1] \"Running binomial downsampling...\"\n",
            "[1] \"9 of 11\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Using depth: 8\"\n",
            "[1] \"Running binomial downsampling...\"\n",
            "[1] \"10 of 11\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Using depth: 16\"\n",
            "[1] \"Running binomial downsampling...\"\n",
            "[1] \"11 of 11\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Using depth: 32\"\n",
            "[1] \"Running binomial downsampling...\"\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FEpEZIZwGEDK",
        "outputId": "bd1bd415-8c08-4a0d-c8e0-30c105ad4246"
      },
      "source": [
        "#############################\n",
        "# mu\n",
        "#############################\n",
        "%%R\n",
        "muVals = 2^seq(-7,3,by=1)\n",
        "mu1 = muVals[length(muVals)] #compare all the others against the highest value\n",
        "\n",
        "muFCUnc = rep(NA,length(muVals))\n",
        "muFCCorr = rep(NA,length(muVals))\n",
        "\n",
        "#just generate one gene with a lot of cells, highly expressed, at two different mu:s\n",
        "\n",
        "g1 = genGeneData(5000, 1000, mu1, totMoleculesPerCell=12000)\n",
        "unc1 = sum(g1[[1]])\n",
        "pred1 = predZTNB(g1[[2]], 10^20)\n",
        "  \n",
        "for(i in 1:length(muVals)) {\n",
        "  print(paste0(i, \" of \", length(muVals)))\n",
        "  \n",
        "  g2 = genGeneData(5000, 1000, muVals[i], totMoleculesPerCell=12000)\n",
        "  unc2 = sum(g2[[1]])\n",
        "  pred2 = predZTNB(g2[[2]], 10^20)\n",
        "  \n",
        "  muFCUnc[i] = abs(log2(unc1/unc2))\n",
        "  muFCCorr[i] = abs(log2(pred1/pred2))\n",
        "}\n",
        "\n",
        "muData = list(muFCUnc = muFCUnc, \n",
        "              muFCCorr = muFCCorr)\n",
        "saveRDS(muData, file =paste0(figure_data_path, \"simMuData.RDS\"))"
      ],
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[1] \"1 of 11\"\n",
            "[1] \"2 of 11\"\n",
            "[1] \"3 of 11\"\n",
            "[1] \"4 of 11\"\n",
            "[1] \"5 of 11\"\n",
            "[1] \"6 of 11\"\n",
            "[1] \"7 of 11\"\n",
            "[1] \"8 of 11\"\n",
            "[1] \"9 of 11\"\n",
            "[1] \"10 of 11\"\n",
            "[1] \"11 of 11\"\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YbJU-zcQGILy",
        "outputId": "5373f609-d9c0-4eb7-c179-7cb56cf88181"
      },
      "source": [
        "#############################\n",
        "# Fold change\n",
        "#############################\n",
        "%%R\n",
        "fcs = 2^(seq(0.5,3,by=(0.5)))\n",
        "fcDE = list()\n",
        "fcFCUnc = list()\n",
        "fcFCCorr = list()\n",
        "\n",
        "for(i in 1:length(fcs)) {\n",
        "  print(paste0(i, \" of \", length(fcs)))\n",
        "  res = evaluateCondition(standardExpr, mus1=mus, mus2=mus, DEExprFactor = fcs[i], numGenes=5000)\n",
        "  fcDE = c(fcDE, list(res[[1]]))\n",
        "  fcFCUnc = c(fcFCUnc, list(res[[2]]))\n",
        "  fcFCCorr = c(fcFCCorr, list(res[[3]]))\n",
        "}\n",
        "\n",
        "fcData = list(de = fcDE,\n",
        "              fcUnc = fcFCUnc,\n",
        "              fcCorr = fcFCCorr)\n",
        "saveRDS(fcData, file =paste0(figure_data_path, \"simFcData.RDS\"))"
      ],
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[1] \"1 of 6\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"2 of 6\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"3 of 6\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"4 of 6\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"5 of 6\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"6 of 6\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gKbg1Ui4GLoS",
        "outputId": "cd115c92-eac0-4284-aaa1-ff9e5870fb13"
      },
      "source": [
        "#############################\n",
        "# Gene expression\n",
        "#############################\n",
        "%%R\n",
        "gexs = 2^(2:13)\n",
        "gexDE = list()\n",
        "gexFCUnc = list()\n",
        "gexFCCorr = list()\n",
        "\n",
        "for(i in 1:length(gexs)) {\n",
        "  print(paste0(i, \" of \", length(gexs)))\n",
        "  res = evaluateCondition(gexs[i], mus1=mus, mus2=mus, DEExprFactor = 2, numGenes=5000)\n",
        "  gexDE = c(gexDE, list(res[[1]]))\n",
        "  gexFCUnc = c(gexFCUnc, list(res[[2]]))\n",
        "  gexFCCorr = c(gexFCCorr, list(res[[3]]))\n",
        "}\n",
        "\n",
        "gexData = list(de = gexDE,\n",
        "               fcUnc = gexFCUnc,\n",
        "               fcCorr = gexFCCorr)\n",
        "saveRDS(gexData, file =paste0(figure_data_path, \"simGexData.RDS\"))"
      ],
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[1] \"1 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"2 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"3 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"4 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"5 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"6 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"7 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"8 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"9 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"10 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"11 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n",
            "[1] \"12 of 12\"\n",
            "[1] \"Generating data...\"\n",
            "[1] \"Predicting...\"\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "McwkqhRJGPza",
        "outputId": "bc10c9e2-dda4-4a40-ec6c-d18610585bc0"
      },
      "source": [
        "#########################\n",
        "#estimate counts per cell\n",
        "#########################\n",
        "%%R\n",
        "expr = rep(NA,length(mus))\n",
        "for(i in 1:length(mus)) {\n",
        "  expr[i] = mean(genGeneData(500, 1000, mus[i])[[1]])\n",
        "}\n",
        "\n",
        "avgGeneCounts = mean(expr)\n",
        "expGeneCounts = 1000/10^6*12000\n",
        "\n",
        "countsPerCell = 12000*avgGeneCounts/expGeneCounts\n",
        "print(countsPerCell)"
      ],
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[1] 7396.055\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Cb5DifYYcB6g",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f84ca178-e5b9-42be-cc35-2bae5485c114"
      },
      "source": [
        "!cd figureData && ls -l"
      ],
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "total 2932\n",
            "drwxr-xr-x 2 root root    4096 Jul  1  2020 PBMC_V3_3\n",
            "-rw-r--r-- 1 root root 1128020 Apr  5 14:06 simDepthData.RDS\n",
            "-rw-r--r-- 1 root root  621725 Apr  5 15:17 simFcData.RDS\n",
            "-rw-r--r-- 1 root root 1237177 Apr  5 17:33 simGexData.RDS\n",
            "-rw-r--r-- 1 root root     281 Apr  5 14:07 simMuData.RDS\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}