{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "GenFig5AB.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/pachterlab/GRNP_2020/blob/master/notebooks/figure_generation/GenFig5AB.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dhydD96df06z"
      },
      "source": [
        "**Generates figure 5 A-B**\n",
        "\n",
        "This notebook generates the batch correction figures in figure 5.\n",
        "\n",
        "Steps:\n",
        "1. Download the code and processed data\n",
        "2. Setup the R environment\n",
        "3. Generate the figures\n",
        "\n",
        "The data for this figure is produced by the following notebooks:\n",
        "\n",
        "Processing of FASTQ files with kallisto and bustools:\n",
        "\n",
        "https://github.com/pachterlab/GRNP_2020/blob/master/notebooks/FASTQ_processing/ProcessPBMC_NG.ipynb\n",
        "\n",
        "https://github.com/pachterlab/GRNP_2020/blob/master/notebooks/FASTQ_processing/ProcessPBMC_NG_2.ipynb\n",
        "\n",
        "Preprocessing of BUG files:\n",
        "\n",
        "https://github.com/pachterlab/GRNP_2020/blob/master/notebooks/R_processing/ProcessR_PBMC_NG.ipynb\n",
        "\n",
        "https://github.com/pachterlab/GRNP_2020/blob/master/notebooks/R_processing/ProcessR_PBMC_NG_2.ipynb\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "h8RnKVMXgbzr"
      },
      "source": [
        "**1. Download the code and processed data**"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "doUAtCxIyOiI",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "deb89c80-b674-404b-fdac-b1dd4d5ac11a"
      },
      "source": [
        "#download the R code\n",
        "![ -d \"GRNP_2020\" ] && rm -r GRNP_2020\n",
        "\n",
        "!git clone https://github.com/pachterlab/GRNP_2020.git\n"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Cloning into 'GRNP_2020'...\n",
            "remote: Enumerating objects: 344, done.\u001b[K\n",
            "remote: Counting objects: 100% (344/344), done.\u001b[K\n",
            "remote: Compressing objects: 100% (284/284), done.\u001b[K\n",
            "remote: Total 2062 (delta 253), reused 87 (delta 60), pack-reused 1718\u001b[K\n",
            "Receiving objects: 100% (2062/2062), 10.91 MiB | 22.03 MiB/s, done.\n",
            "Resolving deltas: 100% (1438/1438), done.\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bJtHkSHB-SSf",
        "outputId": "bfa1d1ff-0ad4-4468-ad34-582a6fd19738"
      },
      "source": [
        "#download BUG data from Zenodo\n",
        "!mkdir data\n",
        "!cd data && wget https://zenodo.org/record/4661569/files/PBMC_NG.zip?download=1 && unzip 'PBMC_NG.zip?download=1' && rm 'PBMC_NG.zip?download=1'"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-04-04 20:06:58--  https://zenodo.org/record/4661569/files/PBMC_NG.zip?download=1\n",
            "Resolving zenodo.org (zenodo.org)... 137.138.76.77\n",
            "Connecting to zenodo.org (zenodo.org)|137.138.76.77|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 502725936 (479M) [application/octet-stream]\n",
            "Saving to: ‘PBMC_NG.zip?download=1’\n",
            "\n",
            "PBMC_NG.zip?downloa 100%[===================>] 479.44M  6.65MB/s    in 61s     \n",
            "\n",
            "2021-04-04 20:08:01 (7.82 MB/s) - ‘PBMC_NG.zip?download=1’ saved [502725936/502725936]\n",
            "\n",
            "Archive:  PBMC_NG.zip?download=1\n",
            "   creating: PBMC_NG/\n",
            "   creating: PBMC_NG/bus_output/\n",
            "  inflating: PBMC_NG/bus_output/bug.txt  \n",
            "  inflating: PBMC_NG/bus_output/coll.genes.txt  \n",
            "   creating: PBMC_NG/bus_output/genecounts/\n",
            "  inflating: PBMC_NG/bus_output/genecounts/output.barcodes.txt  \n",
            "  inflating: PBMC_NG/bus_output/genecounts/output.genes.txt  \n",
            "  inflating: PBMC_NG/bus_output/genecounts/output.mtx  \n",
            "  inflating: PBMC_NG/bus_output/genecounts/output_pred.mtx  \n",
            "  inflating: PBMC_NG/bus_output/transcripts_to_genes.txt  \n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Aou-YnKT-jmK",
        "outputId": "e518c2a1-3596-4a42-be8e-d9e23afdbb8e"
      },
      "source": [
        "!cd data && wget https://zenodo.org/record/4661569/files/PBMC_NG_2.zip?download=1 && unzip 'PBMC_NG_2.zip?download=1' && rm 'PBMC_NG_2.zip?download=1'"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-04-04 20:08:22--  https://zenodo.org/record/4661569/files/PBMC_NG_2.zip?download=1\n",
            "Resolving zenodo.org (zenodo.org)... 137.138.76.77\n",
            "Connecting to zenodo.org (zenodo.org)|137.138.76.77|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 529120611 (505M) [application/octet-stream]\n",
            "Saving to: ‘PBMC_NG_2.zip?download=1’\n",
            "\n",
            "PBMC_NG_2.zip?downl 100%[===================>] 504.61M  8.06MB/s    in 56s     \n",
            "\n",
            "2021-04-04 20:09:20 (8.99 MB/s) - ‘PBMC_NG_2.zip?download=1’ saved [529120611/529120611]\n",
            "\n",
            "Archive:  PBMC_NG_2.zip?download=1\n",
            "   creating: PBMC_NG_2/\n",
            "   creating: PBMC_NG_2/bus_output/\n",
            "  inflating: PBMC_NG_2/bus_output/bug.txt  \n",
            "  inflating: PBMC_NG_2/bus_output/coll.genes.txt  \n",
            "   creating: PBMC_NG_2/bus_output/genecounts/\n",
            "  inflating: PBMC_NG_2/bus_output/genecounts/output.barcodes.txt  \n",
            "  inflating: PBMC_NG_2/bus_output/genecounts/output.genes.txt  \n",
            "  inflating: PBMC_NG_2/bus_output/genecounts/output.mtx  \n",
            "  inflating: PBMC_NG_2/bus_output/genecounts/output_pred.mtx  \n",
            "  inflating: PBMC_NG_2/bus_output/transcripts_to_genes.txt  \n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dUNSQ1qBZb2g",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d3eceb61-6c00-4d27-98ec-8fe376a788ae"
      },
      "source": [
        "#download processed data from Zenodo\n",
        "![ -d \"figureData\" ] && rm -r figureData\n",
        "!mkdir figureData\n",
        "!cd figureData && wget https://zenodo.org/record/4661263/files/PBMC_NG.zip?download=1 && unzip 'PBMC_NG.zip?download=1' && rm 'PBMC_NG.zip?download=1'\n",
        "\n",
        "!cd figureData && wget https://zenodo.org/record/4661263/files/PBMC_NG_2.zip?download=1 && unzip 'PBMC_NG_2.zip?download=1' && rm 'PBMC_NG_2.zip?download=1'\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-04-04 20:09:45--  https://zenodo.org/record/4661263/files/PBMC_NG.zip?download=1\n",
            "Resolving zenodo.org (zenodo.org)... 137.138.76.77\n",
            "Connecting to zenodo.org (zenodo.org)|137.138.76.77|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1183970256 (1.1G) [application/octet-stream]\n",
            "Saving to: ‘PBMC_NG.zip?download=1’\n",
            "\n",
            "PBMC_NG.zip?downloa  14%[=>                  ] 164.79M  9.30MB/s    eta 84s    "
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "oesgTqLO0Qje"
      },
      "source": [
        "#Check that download worked\n",
        "!cd figureData && ls -l && cd PBMC_NG && ls -l"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sCmhNVdYgkWH"
      },
      "source": [
        "**2. Prepare the R environment**"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5Gt6rQkSXriM"
      },
      "source": [
        "#switch to R mode\n",
        "%reload_ext rpy2.ipython\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "jJ3rQJCdgeJa"
      },
      "source": [
        "#install the R packages and setup paths\n",
        "%%R\n",
        "install.packages(\"dplyr\")\n",
        "install.packages(\"textTinyR\")\n",
        "install.packages(\"DescTools\")\n",
        "install.packages(\"qdapTools\")\n",
        "install.packages(\"matrixStats\")\n",
        "install.packages(\"Matrix.utils\")\n",
        "install.packages(\"ggplot2\")\n",
        "install.packages(\"ggpubr\")\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "m_eIE6pR-get"
      },
      "source": [
        "%%R\n",
        "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "\n",
        "BiocManager::install(\"BUSpaRse\", update=FALSE)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "OLVorH03-hTx"
      },
      "source": [
        "%%R\n",
        "install.packages(\"Seurat\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cuzV7hIe92x8"
      },
      "source": [
        "%%R\n",
        "install.packages(\"preseqR\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "x56fjfCSicrp"
      },
      "source": [
        "**3. Generate the figures**\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "V37XLBAO68oR"
      },
      "source": [
        "#First set some path variables\n",
        "%%R\n",
        "source(\"GRNP_2020/RCode/pathsGoogleColab.R\")\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "R6kuhOmzZL_X"
      },
      "source": [
        "#Import helpers (available in other notebooks)\n",
        "%%R\n",
        "source(paste0(sourcePath,\"ButterflyHelpers.R\"))\n",
        "source(paste0(sourcePath,\"CCCHelpers.R\"))\n",
        "source(paste0(sourcePath,\"preseqHelpers.R\"))\n",
        "source(paste0(sourcePath,\"ggplotHelpers.R\"))\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "kqZPO7XjtPPa"
      },
      "source": [
        "#create figure directory\n",
        "![ -d \"figures\" ] && rm -r figures\n",
        "!mkdir figures"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "QzC2G9ZKUlUT"
      },
      "source": [
        "########################\n",
        "#load libraries and data\n",
        "########################\n",
        "%%R\n",
        "library(ggplot2)\n",
        "library(\"ggpubr\")\n",
        "library(tidyverse)\n",
        "library(BUSpaRse)\n",
        "library(textTinyR)\n",
        "library(qdapTools)\n",
        "library(\"matrixStats\")\n",
        "library(Matrix.utils)\n",
        "library(Seurat)\n",
        "\n",
        "colors = gg_color_hue(2)\n",
        "\n",
        "\n",
        "################################################################\n",
        "## Load the PBMC_NG + and PBMC_NG_2 datasets\n",
        "################################################################\n",
        "\n",
        "#remember that the genes file is normally missing in the predquant output folder. Just copy the genes file output by collapse into the folder and rename it.\n",
        "\n",
        "#read NG data\n",
        "#############\n",
        "NGMatrix = read_count_output(paste0(dataPath, \"PBMC_NG/bus_output/genecounts\"), \"output\", tcc=F)\n",
        "#filter cells with less than 200 reads\n",
        "cs = sparse_Sums(NGMatrix, rowSums = F)\n",
        "\n",
        "NGMatrix = NGMatrix[,cs > 200]\n",
        "\n",
        "#convert the genes to gene symbols\n",
        "tr2g = read.table(paste0(dataPath, \"PBMC_NG/bus_output/transcripts_to_genes.txt\"), stringsAsFactors = F)\n",
        "lookupTable = tr2g[,2:3]\n",
        "lookupTable= unique(lookupTable)\n",
        "\n",
        "outGenes = lookup(row.names(NGMatrix), lookupTable)\n",
        "\n",
        "NGMat = NGMatrix\n",
        "row.names(NGMat) = outGenes\n",
        "NGMat = aggregate.Matrix(NGMat, outGenes, fun='sum')\n",
        "\n",
        "#Read NG_2 data\n",
        "###############\n",
        "NG_2Matrix = read_count_output(paste0(dataPath, \"PBMC_NG_2/bus_output/genecounts\"), \"output\", tcc=F)\n",
        "#filter cells with less than 200 reads\n",
        "cs = sparse_Sums(NG_2Matrix, rowSums = F)\n",
        "\n",
        "NG_2Matrix = NG_2Matrix[,cs > 200]\n",
        "\n",
        "#convert the genes to gene symbols\n",
        "tr2g = read.table(paste0(dataPath, \"PBMC_NG_2/bus_output/transcripts_to_genes.txt\"), stringsAsFactors = F)\n",
        "lookupTable = tr2g[,2:3]\n",
        "lookupTable= unique(lookupTable)\n",
        "\n",
        "outGenes = lookup(row.names(NG_2Matrix), lookupTable)\n",
        "\n",
        "NG_2Mat = NG_2Matrix\n",
        "row.names(NG_2Mat) = outGenes\n",
        "NG_2Mat = aggregate.Matrix(NG_2Mat, outGenes, fun='sum')\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "D9NQ-PDrXRh8"
      },
      "source": [
        "%%R\n",
        "############################################\n",
        "#Join the two datasets and process in Seurat\n",
        "############################################\n",
        "#now, test to merge (inner_join) the two sparse matrices\n",
        "mergedmat = merge(NGMat, NG_2Mat, row.names(NGMat), row.names(NG_2Mat), all.x = F, all.y = F)\n",
        "\n",
        "#verify that the merge seemed to work\n",
        "sum(mergedmat[row.names(mergedmat) == \"A2M\",] == c(NGMat[row.names(NGMat) == \"A2M\",],NG_2Mat[row.names(NG_2Mat) == \"A2M\",]))\n",
        "dim(mergedmat)#seems to have worked for that gene at least\n",
        "\n",
        "#now, investigate the matrix in Seurat\n",
        "#########################################\n",
        "\n",
        "set.seed(1)\n",
        "\n",
        "seurObj = CreateSeuratObject(counts = mergedmat, project = \"merged\", min.cells = 0, min.features = 0)\n",
        "seurObj[[\"ds_source\"]] = factor(as.numeric(c(rep(0, dim(NGMat)[2]), rep(1, dim(NG_2Mat)[2]))), c(0,1), c(\"PBMC_NG\",\"PBMC_NG_2\"))\n",
        "seurObj[[\"percent.mt\"]] <- PercentageFeatureSet(seurObj, pattern = \"^MT-\")\n",
        "VlnPlot(seurObj, features = c(\"percent.mt\"))\n",
        "\n",
        "cellFilter = seurObj$percent.mt < 15\n",
        "sourceFilt1 = seurObj$ds_source[cellFilter]\n",
        "mergedmatFilt1 = mergedmat[,cellFilter] #cells already filtered at counts >= 200\n",
        "\n",
        "#easiest to recreate the Seurat object:\n",
        "seurObj = CreateSeuratObject(counts = mergedmatFilt1, project = \"merged\", min.cells = 0, min.features = 0)\n",
        "seurObj[[\"ds_source\"]] = sourceFilt1\n",
        "seurObj <- NormalizeData(seurObj, normalization.method = \"LogNormalize\", scale.factor = 10000)\n",
        "seurObj <- FindVariableFeatures(seurObj, selection.method = \"vst\", nfeatures = 2000)\n",
        "all.genes <- rownames(seurObj)\n",
        "seurObj <- ScaleData(seurObj, features = all.genes)\n",
        "seurObj <- RunPCA(seurObj, features = VariableFeatures(object = seurObj))\n",
        "seurObj <- FindNeighbors(seurObj, dims = 1:10)\n",
        "seurObj <- FindClusters(seurObj, resolution = 0.5)\n",
        "seurObj <- RunUMAP(seurObj, dims = 1:10)\n",
        "\n",
        "#print(FeaturePlot(seurObj, features = c(\"CD19\", \"CD8A\", \"CD3D\"))) # The clusters are T cells\n",
        "\n",
        "clusters = seurObj$seurat_clusters\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Q0yoXUE5Xf8Z"
      },
      "source": [
        "%%R\n",
        "#############################################\n",
        "#Create a subset of the data that better shows the batch effect\n",
        "#############################################\n",
        "subsel = clusters == 0 | clusters == 1 | clusters == 4 | clusters == 5 | clusters == 9\n",
        "dsSource = seurObj$ds_source[subsel]\n",
        "\n",
        "#now the reduced set\n",
        "\n",
        "mergedmatsub = mergedmatFilt1[,subsel]\n",
        "rm(seurObj)\n",
        "gc()\n",
        "\n",
        "seurObj = CreateSeuratObject(counts = mergedmatsub, project = \"merged\", min.cells = 0, min.features = 0)\n",
        "seurObj[[\"ds_source\"]] <- dsSource\n",
        "\n",
        "seurObj <- NormalizeData(seurObj, normalization.method = \"LogNormalize\", scale.factor = 10000)\n",
        "seurObj <- FindVariableFeatures(seurObj, selection.method = \"vst\", nfeatures = 2000)\n",
        "all.genes <- rownames(seurObj)\n",
        "seurObj <- ScaleData(seurObj, features = all.genes)\n",
        "seurObj <- RunPCA(seurObj, features = VariableFeatures(object = seurObj))\n",
        "seurObj <- FindNeighbors(seurObj, dims = 1:10)\n",
        "seurObj <- FindClusters(seurObj, resolution = 0.5)\n",
        "seurObj <- RunUMAP(seurObj, dims = 1:10)\n",
        "pA = DimPlot(seurObj, reduction = \"umap\", group.by = \"ds_source\", cols=c(\"blue\", \"red\"))\n",
        "pA = pA + ggtitle(\"Batch correction - uncorrected\")\n",
        "pA = pA +   theme(panel.background = element_rect(\"white\", \"white\", 0, 0, \"white\"),\n",
        "                  legend.position= \"bottom\", legend.direction = \"horizontal\",#, legend.title = element_blank())\n",
        "                  strip.text.x = element_text(size = 12, face = \"bold\"),\n",
        "                  plot.title = element_text(size=14, face = \"bold\"),\n",
        "                  strip.background = element_blank())\n",
        "pA\n",
        "\n",
        "\n",
        "#extract two things for calculations - the UMAP coordinates and the source\n",
        "uncorrCoords = seurObj[[\"umap\"]]@cell.embeddings\n",
        "uncorrSource = dsSource\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "tkQeT362BM7V"
      },
      "source": [
        "#Create and save the figures\n",
        "\n",
        "\n",
        "set.seed(1)\n",
        "\n",
        "\n",
        "#run through Seurat\n",
        "d = CreateSeuratObject(counts = gn10x, project = \"10x\", min.cells = 0, min.features = 0)\n",
        "\n",
        "#create variable for mitochondrial count fraction\n",
        "d[[\"percent.mt\"]] <- PercentageFeatureSet(d, pattern = \"^MT-\")\n",
        "VlnPlot(d, features = c(\"percent.mt\"))\n",
        "\n",
        "cellFilter = d$percent.mt < 10\n",
        "\n",
        "#easisest to recreate the Seurat object\n",
        "\n",
        "gn10x = gn10x[,cellFilter]\n",
        "gn10xRawCounts = gn10xRawCounts[,cellFilter]\n",
        "d = CreateSeuratObject(counts = gn10x, project = \"10x\", min.cells = 0, min.features = 0)\n",
        "\n",
        "\n",
        "#library size normalization + change to log scale\n",
        "d <- NormalizeData(d, normalization.method = \"LogNormalize\", scale.factor = 10000)\n",
        "\n",
        "#Finds the most variable genes\n",
        "d <- FindVariableFeatures(d, selection.method = \"vst\", nfeatures = 2000)\n",
        "\n",
        "#Make mean and variance the same for all genes:\n",
        "d <- ScaleData(d, features = rownames(d))\n",
        "\n",
        "#Principal component analysis\n",
        "d <- RunPCA(d, features = VariableFeatures(object = d))\n",
        "\n",
        "#do clustering\n",
        "d <- FindNeighbors(d, dims = 1:10)\n",
        "d <- FindClusters(d, resolution = 0.5)\n",
        "\n",
        "#Generate UMAP map\n",
        "d <- RunUMAP(d, dims = 1:10)\n",
        "\n",
        "clust = d$seurat_clusters\n",
        "\n",
        "\n",
        "d = RenameIdents(object = d, `0` = \"T1\", `1` = \"T2\", `2` = \"T3\", `3` = \"M1\", `4` = \"B\", `5` = \"M2\", `6` = \"T4\", `7` = \"T5\", `8` = \"U1\", `9` = \"U2\", `10` = \"U3\",  `11` = \"U4\")\n",
        "\n",
        "#Test 1 - make sure we identified the clusters correctly\n",
        "d2 = subset(d, seurat_clusters == 0 | seurat_clusters == 1 | seurat_clusters == 2 | seurat_clusters == 6  | seurat_clusters == 7 ) #T/NK cells\n",
        "DimPlot(d2, reduction = \"umap\")\n",
        "d2 = subset(d, seurat_clusters == 3 | seurat_clusters == 5 ) #Monocytes\n",
        "DimPlot(d2, reduction = \"umap\")\n",
        "d2 = subset(d, seurat_clusters == 4) #B cells\n",
        "DimPlot(d2, reduction = \"umap\")\n",
        "#Test 2 - Identify the cell types in the clusters\n",
        "FeaturePlot(d, c(\"CD3D\", \"CD19\", \"LYZ\"))\n",
        "            \n",
        "\n",
        "#show clustering\n",
        "pSupA = DimPlot(d, reduction = \"umap\")\n",
        "print(pSupA)\n",
        "\n",
        "\n",
        "ggsave(\n",
        "  paste0(figure_path, \"FigS24.png\"),\n",
        "  plot = pSupA, device = \"png\",\n",
        "  width = 4, height = 4, dpi = 300)\n",
        "\n",
        "\n",
        "\n",
        "######################\n",
        "# Compare CU across clusters\n",
        "######################\n",
        "\n",
        "cuPerCluster = matrix(, nrow = nrow(gn10x), ncol = 12) #number of clusters is 12\n",
        "rownames(cuPerCluster) = rownames(gn10x)\n",
        "UMIsPerCluster = matrix(, nrow = nrow(gn10x), ncol = 12) #number of clusters is 12\n",
        "rownames(UMIsPerCluster) = rownames(gn10x)\n",
        "cellsPerCluster = rep(NA,12)\n",
        "clusterNames = c(\"T1\",\"T2\",\"T3\",\"M1\",\"B\",\"M2\",\"T4\",\"T5\",\"U1\",\"U2\",\"U3\",\"U4\") #compared manually with id setting above\n",
        "colnames(cuPerCluster) = clusterNames\n",
        "colnames(UMIsPerCluster) = clusterNames\n",
        "\n",
        "#load the bug\n",
        "loadBug(\"EVALPBMC\")\n",
        "bug = getBug(\"EVALPBMC\")\n",
        "\n",
        "for (i in 1:12) {\n",
        "  sel = d$seurat_clusters == i-1\n",
        "  cellsPerCluster[i] = sum(sel)\n",
        "  UMIsPerCluster[,i] = sparse_Sums(gn10x[,sel], rowSums = TRUE)\n",
        "\n",
        "  #to figure out cu per cluster, we need to use the bug\n",
        "  barcodes = colnames(gn10x)[sel]\n",
        "  subBug = bug[bug$barcode %in% barcodes,]\n",
        "  \n",
        "  res = subBug %>% group_by(gene) %>% summarize(cu = mean(count))\n",
        "  colnames(res) = c(\"x\",\"y\")\n",
        "  cuPerCluster[,i] = lookup(rownames(cuPerCluster), res)\n",
        "}\n",
        "\n",
        "\n",
        "cuPerClusterFilt = cuPerCluster[,cellsPerCluster > 300]\n",
        "UMIsPerClusterFilt = UMIsPerCluster[,cellsPerCluster > 300]\n",
        "cuPerClusterFilt[UMIsPerClusterFilt < 20] = NA\n",
        "\n",
        "#only measure the variance for somewhat highly expressed genes\n",
        "cuPerClusterFilt2 = cuPerClusterFilt[(rowMeans(UMIsPerClusterFilt) > 60) & (rowSums(!is.na(cuPerClusterFilt)) > 2),]\n",
        "dim(cuPerClusterFilt2)\n",
        "cuPerClusterFilt2\n",
        "variances = rowVars(cuPerClusterFilt2, na.rm=TRUE)\n",
        "\n",
        "srt = sort(variances, index.return=TRUE, decreasing = TRUE)\n",
        "sel = seq_len(length(srt$x)) %in% srt$ix[1:50] #get the 50 genes with most variance\n",
        "\n",
        "#sel = variances > 10\n",
        "sum(sel)\n",
        "\n",
        "TPMs = UMIsPerCluster\n",
        "for (i in 1:12) {\n",
        "  TPMs[,i] = TPMs[,i]*10^6/sum(TPMs[,i])\n",
        "}\n",
        "colSums(TPMs)#ok\n",
        "\n",
        "\n",
        "cuForVariableGenes = cuPerClusterFilt2[sel,]\n",
        "UMIsForVariableGenes = UMIsPerCluster[rownames(UMIsPerCluster) %in% rownames(cuForVariableGenes),]\n",
        "TPMsForVariableGenes = TPMs[rownames(TPMs) %in% rownames(cuForVariableGenes),cellsPerCluster > 300]\n",
        "logtrans = log2(TPMsForVariableGenes + 1)\n",
        "\n",
        "max(variances, na.rm=TRUE)\n",
        "\n",
        "geneSel = rownames(cuForVariableGenes) == \"ALDH2\" #ALDH2\n",
        "cu = cuForVariableGenes[geneSel,]\n",
        "logExpr = logtrans[geneSel,]\n",
        "texts = names(cu)\n",
        "\n",
        "dsPlot = tibble(x=cu, y=logExpr, col=c(colors[1],colors[1],colors[1],colors[2],\"#000000\",colors[2],colors[1],colors[1]))\n",
        "dsText = tibble(x=cu + 0.2, y=logExpr)\n",
        "fit = lm(logExpr~cu)\n",
        "summary(fit) #p = 0.0040, F-Test\n",
        "#fix the texts a bit so they are visible and don't overlap\n",
        "dsText$x[6] = dsText$x[6] - 1.5\n",
        "dsText$y[8] = dsText$y[8] + 0.05\n",
        "dsText$y[3] = dsText$y[3] - 0.05\n",
        "\n",
        "pE = ggplot2::ggplot(dsPlot,ggplot2::aes(x=x,y=y)) +\n",
        "  ggplot2::geom_point(ggplot2::aes(x=x,y=y), color=dsPlot$col, shape=1, size=1) +\n",
        "  geom_abline(slope = fit$coefficients[2], intercept = fit$coefficients[1], colour=\"#008800\", size=1.3) +\n",
        "  geom_text(data=dsText, label=texts, size=4, hjust = 0, color=dsPlot$col, parse=FALSE) +\n",
        "  ggplot2::labs(y=expression(Log[2]*\"(CPM + 1)\"), x=\"Copies per UMI\", title=\"ALDH2 acr. clusters, uncorr.\") +\n",
        "  theme(panel.background = element_rect(\"white\", \"white\", 0, 0, \"white\"),\n",
        "        legend.position= \"bottom\", legend.direction = \"horizontal\",#, legend.title = element_blank())\n",
        "        strip.text.x = element_text(size = 12, face = \"bold\"),\n",
        "        #legend.position= \"none\",\n",
        "        plot.title = element_text(face = \"bold\"),\n",
        "        strip.background = element_blank())\n",
        "print(pE)\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "###########################\n",
        "#look at the histograms for T cells and Monocytes\n",
        "###########################\n",
        "\n",
        "#first extract the barcodes\n",
        "sel = d$seurat_clusters == 0 | d$seurat_clusters == 1 | d$seurat_clusters == 2 | d$seurat_clusters == 6 | d$seurat_clusters == 7\n",
        "tcellBarcodes = colnames(gn10x)[sel]\n",
        "sel = d$seurat_clusters == 3 | d$seurat_clusters == 5\n",
        "monoBarcodes = colnames(gn10x)[sel]\n",
        "\n",
        "#load the bug\n",
        "tCounts = bug[(bug$barcode %in% tcellBarcodes) & (bug$gene == \"ALDH2\"),]$count\n",
        "mCounts = bug[(bug$barcode %in% monoBarcodes) & (bug$gene == \"ALDH2\"),]$count\n",
        "mCountsFilt = mCounts[mCounts < 30]\n",
        "hend = max(c(mCountsFilt,tCounts))\n",
        "numTTotMolecules = nrow(bug[bug$barcode %in% tcellBarcodes,])\n",
        "numMTotMolecules = nrow(bug[bug$barcode %in% monoBarcodes,])\n",
        "numTTotCounts = sum(bug[bug$barcode %in% tcellBarcodes,]$count)\n",
        "numMTotCounts = sum(bug[bug$barcode %in% monoBarcodes,]$count)\n",
        "\n",
        "\n",
        "ht = hist(tCounts, breaks=seq(0.5, hend+0.5, by=1), plot = FALSE)$counts\n",
        "hm = hist(mCountsFilt, breaks=seq(0.5, hend+0.5, by=1), plot = FALSE)$counts\n",
        "\n",
        "htScaled = ht/numTTotMolecules * 10^6\n",
        "hmScaled = hm/numMTotMolecules * 10^6\n",
        "\n",
        "\n",
        "dsPlot = data.frame(x = 1:hend, y = htScaled)\n",
        "pC = ggplot(dsPlot,aes(x=x,y=y)) +\n",
        "  geom_bar(stat=\"identity\", fill = colors[1]) +\n",
        "  labs(y=\"CPM\", x=\"Counts per UMI\", title=\"ALDH2, T cells\") +\n",
        "  ylim(0,50) +\n",
        "  theme(panel.background = element_rect(\"white\", \"white\", 0, \n",
        "                                        0, \"white\"),\n",
        "        plot.title = element_text(face = \"bold\")\n",
        "  )\n",
        "print(pC)\n",
        "\n",
        "dsPlot = data.frame(x = 1:hend, y = hmScaled)\n",
        "pD = ggplot(dsPlot,aes(x=x,y=y)) +\n",
        "  geom_bar(stat=\"identity\", fill = colors[2]) +\n",
        "  labs(y=\"CPM\", x=\"Counts per UMI\", title=\"ALDH2, Monocytes\") +\n",
        "  ylim(0,50) +\n",
        "  theme(panel.background = element_rect(\"white\", \"white\", 0, \n",
        "                                        0, \"white\"),\n",
        "        plot.title = element_text(face = \"bold\")\n",
        "  )\n",
        "print(pD)\n",
        "\n",
        "\n",
        "##############################\n",
        "#Plot all variable genes\n",
        "##############################\n",
        "\n",
        "#first scale all genes to the same scale (sum of 1000)\n",
        "scaled = TPMsForVariableGenes*1000/rowSums(TPMsForVariableGenes)\n",
        "rowSums(scaled)\n",
        "logScaled = log2(scaled+1)\n",
        "cu = NULL\n",
        "scLogExpr = NULL\n",
        "\n",
        "for (i in 1:nrow(cuForVariableGenes)) {\n",
        "  sel = !is.na(cuForVariableGenes[i,])\n",
        "  cu = c(cu, as.numeric(cuForVariableGenes[i,sel]))  \n",
        "  scLogExpr = as.numeric(c(scLogExpr, logScaled[i,sel]))\n",
        "}\n",
        "\n",
        "fit = lm(scLogExpr~cu)\n",
        "summary(fit)# p-value: < 2.2e-16, F-Test\n",
        "\n",
        "dsPlot = tibble(x=cu, y=scLogExpr)\n",
        "dsText = tibble(x=1, y=10)\n",
        "\n",
        "\n",
        "pG = ggplot2::ggplot(dsPlot,ggplot2::aes(x=x,y=y)) +\n",
        "  ggplot2::geom_point(ggplot2::aes(x=x,y=y), color=\"black\", shape=1, size=1) +\n",
        "  geom_abline(slope = fit$coefficients[2], intercept = fit$coefficients[1], colour=\"#008800\", size=1.3) +\n",
        "  geom_text(data=dsText, label=paste0(\"R = \",format(cor(cu,scLogExpr), digits=2)), size=5, hjust = 0, parse=FALSE) +\n",
        "  ggplot2::labs(y=expression(Log[2]*\"(Norm. Expr + pc)\"), x=\"Copies per UMI\", title=\"Ampl. acr.clusters, uncorr.\") +\n",
        "  theme(panel.background = element_rect(\"white\", \"white\", 0, 0, \"white\"),\n",
        "        legend.position= \"bottom\", legend.direction = \"horizontal\",#, legend.title = element_blank())\n",
        "        strip.text.x = element_text(size = 12, face = \"bold\"),\n",
        "        plot.title = element_text(face = \"bold\"),\n",
        "        strip.background = element_blank())\n",
        "print(pG)\n",
        "\n",
        "\n",
        "######################\n",
        "# Now predict and see if the correlation goes down\n",
        "######################\n",
        "\n",
        "dsid = \"EVALPBMC\"\n",
        "bug = getBug(dsid)\n",
        "\n",
        "#first filter on the genes in the plot to make the bug smaller\n",
        "genes = rownames(cuForVariableGenes)\n",
        "bugFilt1 = bug[bug$gene %in% genes,]\n",
        "\n",
        "predDS = UMIsForVariableGenes #just allocate the right size\n",
        "predDS[,] = NA\n",
        "predZTNB = UMIsForVariableGenes #just allocate the right size\n",
        "predZTNB[,] = NA\n",
        "\n",
        "\n",
        "for (i in 1:12) {\n",
        "  print(i)\n",
        "  sel = clust == i-1\n",
        "  cellIds = names(clust)[sel]\n",
        "  clustBug = bugFilt1[bugFilt1$barcode %in% cellIds,] #filter a bit more\n",
        "  if (cellsPerCluster[i] > 300) {\n",
        "    for (g in 1:length(genes)) {\n",
        "      if (UMIsForVariableGenes[g,i] >= 20) {\n",
        "        #Get the histogram\n",
        "        counts = clustBug[clustBug$gene == genes[g],]$count\n",
        "        h = hist(counts, breaks=seq(0.5, max(counts)+0.5, by=1), plot = FALSE)\n",
        "        freq = h$mids\n",
        "        counts = h$counts\n",
        "        added = 0\n",
        "        #preseq cannot handle if we have only ones, so modify the histogram slightly\n",
        "        if ((length(freq)==1) & (freq[1] == 1)) {\n",
        "          added = 2\n",
        "          freq = c(1,2)\n",
        "          counts = c(counts[1]+1,1)#room for improvement here\n",
        "        }\n",
        "        dd = as.matrix(data.frame(freq,counts));\n",
        "        rSACZTNB = mod.ztnb.rSAC(dd, incTol = 1e-5, iterIncTol = 200);\n",
        "        rSACDS = ds.rSAC(dd,mt=2)\n",
        "        newCountsZTNB = rSACZTNB(10^20)\n",
        "        newCountsDS = rSACDS(10^20)\n",
        "        newCountsZTNB[newCountsZTNB < 0] = 0\n",
        "        newCountsDS[newCountsDS < 0] = 0\n",
        "        predZTNB[g,i] = newCountsZTNB - added;\n",
        "        predDS[g,i] = newCountsDS - added;\n",
        "      }\n",
        "    }\n",
        "  }\n",
        "}\n",
        "\n",
        "\n",
        "#remove all columns with just NA\n",
        "predFiltZTNB = predZTNB[,colSums(!is.na(predZTNB)) != 0]\n",
        "predLogZTNB = predFiltZTNB;\n",
        "predFiltDS = predDS[,colSums(!is.na(predDS)) != 0]\n",
        "predLogDS = predFiltDS;\n",
        "#scale to the same number of average counts per cluster as before\n",
        "colSumsData = TPMsForVariableGenes\n",
        "colSumsData[is.na(predFiltZTNB)] = NA\n",
        "scaleSizeZTNB = mean(colSums(colSumsData,na.rm=TRUE))\n",
        "colSumsData = TPMsForVariableGenes\n",
        "colSumsData[is.na(predFiltDS)] = NA\n",
        "scaleSizeDS = mean(colSums(colSumsData,na.rm=TRUE))\n",
        "\n",
        "\n",
        "for (i in 1:ncol(predFiltZTNB)) {\n",
        "  TPMish = predFiltZTNB[,i]*scaleSizeZTNB/sum(predFiltZTNB, na.rm = TRUE)\n",
        "  predLogZTNB[,i] = log2(TPMish + 1)\n",
        "  TPMish = predFiltDS[,i]*scaleSizeDS/sum(predFiltDS, na.rm = TRUE)\n",
        "  predLogDS[,i] = log2(TPMish + 1)\n",
        "}\n",
        "\n",
        "#use DS in main figure\n",
        "\n",
        "#Get the average TPM per cluster before prediction\n",
        "sumTPMBefore = sum(TPMsForVariableGenes[rownames(TPMsForVariableGenes) == \"ALDH2\",])\n",
        "\n",
        "geneSel = rownames(predFiltDS) == \"ALDH2\" #ALDH2\n",
        "cu = cuForVariableGenes[geneSel,]\n",
        "pseudoTPM = predFiltDS[geneSel,] * sumTPMBefore/sum(predFiltDS[geneSel,])\n",
        "logExpr = log2(pseudoTPM + 1)\n",
        "fit = lm(logExpr~cu)\n",
        "summary(fit) # p value: 0.5355, F-Test\n",
        "\n",
        "dsPlot = tibble(x=cu, y=logExpr, col=c(colors[1],colors[1],colors[1],colors[2],\"#000000\",colors[2],colors[1],colors[1]))\n",
        "dsText = tibble(x=cu + 0.2, y=logExpr)\n",
        "#fix the texts a bit so they are visible and don't overlap\n",
        "dsText$x[6] = dsText$x[6] - 1.5\n",
        "dsText$y[1] = dsText$y[1] - 0.1\n",
        "\n",
        "pF = ggplot2::ggplot(dsPlot,ggplot2::aes(x=x,y=y)) +\n",
        "  ggplot2::geom_point(ggplot2::aes(x=x,y=y), color=dsPlot$col, shape=1, size=1) +\n",
        "  geom_abline(slope = fit$coefficients[2], intercept = fit$coefficients[1], colour=\"#008800\", size=1.3) +\n",
        "  geom_text(data=dsText, label=texts, size=4, hjust = 0, color=dsPlot$col, parse=FALSE) +\n",
        "  ggplot2::labs(y=expression(Log[2]*\"(pseudoTPM + pc)\"), x=\"Copies per UMI\", title=\"ALDH2 acr. clusters, corr.\") +\n",
        "  theme(panel.background = element_rect(\"white\", \"white\", 0, 0, \"white\"),\n",
        "        legend.position= \"bottom\", legend.direction = \"horizontal\",#, legend.title = element_blank())\n",
        "        strip.text.x = element_text(size = 12, face = \"bold\"),\n",
        "        plot.title = element_text(face = \"bold\"),\n",
        "        strip.background = element_blank())\n",
        "print(pF)\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "##############################\n",
        "#Plot all variable genes, now corrected\n",
        "##############################\n",
        "\n",
        "#first scale all genes to the same scale (sum of 1000)\n",
        "scaled = predFiltDS*1000/rowSums(predFiltDS, na.rm=TRUE)\n",
        "rowSums(scaled, na.rm=TRUE)\n",
        "logScaled = log2(scaled+1)\n",
        "cu = NULL\n",
        "scLogExpr = NULL\n",
        "\n",
        "for (i in 1:nrow(cuForVariableGenes)) {\n",
        "  sel = !is.na(cuForVariableGenes[i,])\n",
        "  cu = c(cu, as.numeric(cuForVariableGenes[i,sel]))  \n",
        "  scLogExpr = as.numeric(c(scLogExpr, logScaled[i,sel]))\n",
        "}\n",
        "\n",
        "fit = lm(scLogExpr~cu)\n",
        "summary(fit)#p = 5.111e-09, F-Test\n",
        "dsPlot = tibble(x=cu, y=scLogExpr)\n",
        "dsText = tibble(x=1, y=10)\n",
        "\n",
        "\n",
        "pH = ggplot2::ggplot(dsPlot,ggplot2::aes(x=x,y=y)) +\n",
        "  ggplot2::geom_point(ggplot2::aes(x=x,y=y), color=\"black\", shape=1, size=1) +\n",
        "  geom_abline(slope = fit$coefficients[2], intercept = fit$coefficients[1], colour=\"#008800\", size=1.3) +\n",
        "  geom_text(data=dsText, label=paste0(\"R = \",format(cor(cu,scLogExpr), digits=2)), size=5, hjust = 0, parse=FALSE) +\n",
        "  ggplot2::labs(y=expression(Log[2]*\"(Norm. Expr + pc)\"), x=\"Copies per UMI\", title=\"Ampl. acr. clusters, corr.\") +\n",
        "  theme(panel.background = element_rect(\"white\", \"white\", 0, 0, \"white\"),\n",
        "        legend.position= \"bottom\", legend.direction = \"horizontal\",#, legend.title = element_blank())\n",
        "        strip.text.x = element_text(size = 12, face = \"bold\"),\n",
        "        plot.title = element_text(face = \"bold\"),\n",
        "        #legend.position= \"none\",\n",
        "        strip.background = element_blank())\n",
        "print(pH)\n",
        "\n",
        "fig5CH = ggarrange(pC, pD, pE, pF, pG, pH, nrow=2, ncol=3, labels=c(\"C\",\"D\",\"E\",\"F\",\"G\",\"H\"))\n",
        "\n",
        "print(fig5CH)\n",
        "\n",
        "ggsave(\n",
        "  paste0(figure_path, \"Fig5C-H.png\"),\n",
        "  plot = fig5CH, device = \"png\",\n",
        "  width = 9, height = 6, dpi = 300)\n",
        "\n",
        "#Create supplementary figure with ZTNB\n",
        "################\n",
        "\n",
        "scaled = predFiltZTNB*1000/rowSums(predFiltZTNB, na.rm=TRUE)\n",
        "rowSums(scaled, na.rm=TRUE)\n",
        "logScaled = log2(scaled+1)\n",
        "cu = NULL\n",
        "scLogExpr = NULL\n",
        "\n",
        "for (i in 1:nrow(cuForVariableGenes)) {\n",
        "  sel = !is.na(cuForVariableGenes[i,])\n",
        "  cu = c(cu, as.numeric(cuForVariableGenes[i,sel]))  \n",
        "  scLogExpr = as.numeric(c(scLogExpr, logScaled[i,sel]))\n",
        "}\n",
        "\n",
        "fit = lm(scLogExpr~cu)\n",
        "summary(fit)#p=0.4397, F-Test\n",
        "dsPlot = tibble(x=cu, y=scLogExpr)\n",
        "dsText = tibble(x=1, y=10)\n",
        "\n",
        "\n",
        "pSup2 = ggplot2::ggplot(dsPlot,ggplot2::aes(x=x,y=y)) +\n",
        "  ggplot2::geom_point(ggplot2::aes(x=x,y=y), color=\"black\", shape=1, size=1) +\n",
        "  geom_abline(slope = fit$coefficients[2], intercept = fit$coefficients[1], colour=\"#008800\", size=1.3) +\n",
        "  geom_text(data=dsText, label=paste0(\"R = \",format(cor(cu,scLogExpr), digits=2)), size=5, hjust = 0, parse=FALSE) +\n",
        "  ggplot2::labs(y=expression(Log[2]*\"(Norm. Expr + pc)\"), x=\"Copies per UMI\") +\n",
        "  theme(panel.background = element_rect(\"white\", \"white\", 0, 0, \"white\"),\n",
        "        legend.position= \"bottom\", legend.direction = \"horizontal\",#, legend.title = element_blank())\n",
        "        strip.text.x = element_text(size = 12, face = \"bold\"),\n",
        "        #legend.position= \"none\",\n",
        "        strip.background = element_blank())\n",
        "print(pSup2)\n",
        "\n",
        "ggsave(\n",
        "  paste0(figure_path, \"FigS25.png\"),\n",
        "  plot = pSup2, device = \"png\",\n",
        "  width = 3, height = 3, dpi = 300)\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}